File: 0719-find-k-th-smallest-pair-distance/0719-find-k-th-smallest-pair-distance.py
class Solution:
    def smallestDistancePair(self, nums: List[int], k: int) -> int:
        # Sort the numbers to make pair distance calculation easier
        nums.sort()
        
        def countPairs(max_distance: int) -> int:
            # This function returns the number of pairs with a distance <= max_distance
            count = 0
            start = 0
            for i in range(len(nums)):
                # Find the rightmost pair (nums[i], nums[j]) such that nums[j] - nums[i] <= max_distance
                while start < len(nums) and nums[i] - nums[start] > max_distance:
                    start += 1
                count += i - start
            return count
        
        # Binary search on the distance
        low, high = 0, nums[-1] - nums[0]
        
        while low < high:
            mid = (low + high) // 2
            if countPairs(mid) < k:
                low = mid + 1
            else:
                high = mid
        
        return low
        

File: C. Hard/0719-find-k-th-smallest-pair-distance/0719-find-k-th-smallest-pair-distance.c
int compare(const void *a, const void *b) {
    return (*(int *)a - *(int *)b);
}

int countPairs(int *nums, int numsSize, int maxDistance) {
    int count = 0;
    int start = 0;
    for (int i = 0; i < numsSize; i++) {
        while (start < numsSize && nums[i] - nums[start] > maxDistance) {
            start++;
        }
        count += i - start;
    }
    return count;
}

int smallestDistancePair(int* nums, int numsSize, int k) {
    // Sort the array
    qsort(nums, numsSize, sizeof(int), compare);
    
    int low = 0;
    int high = nums[numsSize - 1] - nums[0];
    
    while (low < high) {
        int mid = (low + high) / 2;
        if (countPairs(nums, numsSize, mid) < k) {
            low = mid + 1;
        } else {
            high = mid;
        }
    }
    
    return low;
}